File: lib/editor/trace/pam-palette.ts
/**
 * PAM (Partitioning Around Medoids) palette-medoid selection — client side.
 *
 * Mirror of `filter-service/app/pam_palette.py`. Same BUILD + SWAP
 * phases per Kaufman & Rousseeuw (1987); byte-equivalent output against
 * the server (asserted by the same constructed test cases in both
 * `pam-palette.test.ts` and `test_pam_palette.py`).
 *
 * Replacement for the count-based `reduce_to_top_n` palette cap, which
 * keeps the most-frequent snap winners. Count-based reduction is
 * dominant-preserving but spread-unaware: a small saturated region
 * whose chips don't make the top-N gets re-snapped to whatever was
 * popular, losing the colour cluster entirely. PAM is spread-optimal:
 * it minimises total snap distance over the whole image, so rare-but-
 * distinct colour clusters keep a representative.
 *
 * Two phases:
 *   - BUILD: greedy init. First medoid minimises Σ_i D[i, m]; each
 *     subsequent medoid picks the candidate with the largest cost
 *     improvement when added.
 *   - SWAP: local search. For each (current medoid, non-medoid
 *     candidate) pair compute the cost change if we swap; apply the
 *     best improving swap until no improvement exists.
 *
 * Distance-metric agnostic — callers precompute the (N, M) distance
 * matrix `D` using OKLab squared-Euclidean, CIEDE2000, or any other
 * metric and pass it in. PR-C ships PAM as a pure building block; PR-F
 * wires it into the snap pipeline.
 */

/**
 * A flat (N, M) distance matrix stored in row-major order: `D[i*M + j]`
 * is the distance from point `i` to candidate `j`. `N` is `data.length /
 * M` (with `data.length % M === 0` asserted at the boundary).
 */
export type DistanceMatrix = {
  readonly data: Float64Array
  readonly N: number
  readonly M: number
}

/** Build a `DistanceMatrix` from a nested array `D[i][j]`. */
export function distanceMatrixFromRows(rows: ReadonlyArray<ReadonlyArray<number>>): DistanceMatrix {
  const N = rows.length
  const M = N === 0 ? 0 : rows[0].length
  const data = new Float64Array(N * M)
  for (let i = 0; i < N; i += 1) {
    if (rows[i].length !== M) {
      throw new Error(`distanceMatrixFromRows: ragged row ${i} (got ${rows[i].length}, expected ${M})`)
    }
    for (let j = 0; j < M; j += 1) data[i * M + j] = rows[i][j]
  }
  return { data, N, M }
}

function totalCost(D: DistanceMatrix, medoids: ReadonlyArray<number>, weights: ReadonlyArray<number> | null): number {
  let cost = 0
  for (let i = 0; i < D.N; i += 1) {
    let best = Infinity
    const rowBase = i * D.M
    for (let mi = 0; mi < medoids.length; mi += 1) {
      const d = D.data[rowBase + medoids[mi]]
      if (d < best) best = d
    }
    cost += weights === null ? best : best * weights[i]
  }
  return cost
}

function buildPhase(D: DistanceMatrix, k: number, weights: ReadonlyArray<number> | null): number[] {
  // First medoid: column with the smallest weighted-sum distance.
  let bestFirst = 0
  let bestFirstCost = Infinity
  for (let j = 0; j < D.M; j += 1) {
    let col = 0
    for (let i = 0; i < D.N; i += 1) {
      const d = D.data[i * D.M + j]
      col += weights === null ? d : d * weights[i]
    }
    if (col < bestFirstCost) {
      bestFirstCost = col
      bestFirst = j
    }
  }
  const medoids: number[] = [bestFirst]

  // nearest[i] = D[i, current_nearest_medoid]
  const nearest = new Float64Array(D.N)
  for (let i = 0; i < D.N; i += 1) nearest[i] = D.data[i * D.M + bestFirst]

  const inMedoids = new Uint8Array(D.M)
  inMedoids[bestFirst] = 1

  for (let step = 1; step < k; step += 1) {
    let bestGain = -Infinity
    let bestCandidate = -1
    for (let c = 0; c < D.M; c += 1) {
      if (inMedoids[c]) continue
      let gain = 0
      for (let i = 0; i < D.N; i += 1) {
        const d = D.data[i * D.M + c]
        if (d < nearest[i]) {
          const delta = nearest[i] - d
          gain += weights === null ? delta : delta * weights[i]
        }
      }
      if (gain > bestGain) {
        bestGain = gain
        bestCandidate = c
      }
    }
    if (bestCandidate < 0) break // can happen only if M < k (already validated)
    medoids.push(bestCandidate)
    inMedoids[bestCandidate] = 1
    for (let i = 0; i < D.N; i += 1) {
      const d = D.data[i * D.M + bestCandidate]
      if (d < nearest[i]) nearest[i] = d
    }
  }

  return medoids
}

function swapPhase(
  D: DistanceMatrix,
  initial: number[],
  weights: ReadonlyArray<number> | null,
  maxSwaps: number,
): number[] {
  const medoids = [...initial]
  for (let iter = 0; iter < maxSwaps; iter += 1) {
    const currentCost = totalCost(D, medoids, weights)
    let bestDelta = 0
    let bestSwap: { mi: number; c: number } | null = null

    for (let mi = 0; mi < medoids.length; mi += 1) {
      const trial = [...medoids]
      for (let c = 0; c < D.M; c += 1) {
        if (medoids.includes(c)) continue
        trial[mi] = c
        const cost = totalCost(D, trial, weights)
        const delta = cost - currentCost
        if (delta < bestDelta) {
          bestDelta = delta
          bestSwap = { mi, c }
        }
      }
    }
    if (bestSwap === null) break
    medoids[bestSwap.mi] = bestSwap.c
  }
  return medoids
}

/**
 * Pick `k` medoid indices from the candidate axis of `D`. See module
 * doc for the algorithm; tie-breaks via the lowest-index winner of
 * `argmin` (matches the Python `np.argmin` semantics) so PAM is
 * deterministic.
 *
 * @throws when `k` is outside `(0, M]` or `weights.length !== N`.
 */
export function pamSelectMedoids(
  D: DistanceMatrix,
  k: number,
  options: { weights?: ReadonlyArray<number>; maxSwaps?: number } = {},
): number[] {
  if (!(0 < k && k <= D.M)) {
    throw new Error(`pamSelectMedoids: k must satisfy 0 < k <= ${D.M}; got ${k}`)
  }
  const weights = options.weights ?? null
  if (weights !== null && weights.length !== D.N) {
    throw new Error(`pamSelectMedoids: weights.length must equal N (${D.N}); got ${weights.length}`)
  }
  const maxSwaps = options.maxSwaps ?? 100
  const built = buildPhase(D, k, weights)
  const swapped = swapPhase(D, built, weights, maxSwaps)
  return swapped.slice().sort((a, b) => a - b)
}

File: filter-service/app/pam_palette.py
"""
PAM (Partitioning Around Medoids) palette-medoid selection.

Implementation of Kaufman & Rousseeuw (1987), "Clustering by Means of
Medoids" — the standard k-medoid clustering algorithm. Used to pick `k`
chip indices from a palette pool that minimise the total distance from
a set of source points (e.g. cell means) to their nearest chip.

Replacement for the count-based `reduce_to_top_n` (`palette_reduction.py`)
which keeps the most-frequent snap winners. Count-based reduction is
dominant-preserving but spread-unaware: a small saturated region whose
chips don't make the top-N gets re-snapped to whatever was popular,
losing the colour cluster entirely. PAM is spread-optimal: it
minimises total snap distance over the whole image, so rare-but-
distinct colour clusters keep a representative.

Two phases per Kaufman & Rousseeuw:
  - BUILD: greedy init. First medoid minimises Σ_i D[i, m]; each
    subsequent medoid picks the candidate with the largest cost
    improvement when added.
  - SWAP: local search. For each (current medoid, non-medoid candidate)
    pair compute the cost change if we swap; apply the best
    improving swap until no improvement exists.

The implementation is distance-metric agnostic: callers precompute the
(N, M) distance matrix `D` using OKLab squared-Euclidean, CIEDE2000, or
any other metric and pass it in. This keeps PAM purely algorithmic.

A `lib/editor/trace/pam-palette.ts` mirror runs in the client preview;
parity is asserted by the same constructed test cases.

This module is purely additive — no caller wires it into the trace
pipeline yet. PR-F integrates PAM into the snap step; PR-C just
provides the algorithm + parity-tested correctness against known
constructions.
"""
from __future__ import annotations

import numpy as np


def _total_cost(D: np.ndarray, medoids: np.ndarray, weights: np.ndarray | None) -> float:
    """Σ_i w_i · min_j D[i, medoid_j], the PAM objective."""
    nearest = D[:, medoids].min(axis=1)
    if weights is None:
        return float(nearest.sum())
    return float((nearest * weights).sum())


def _build_phase(
    D: np.ndarray, k: int, weights: np.ndarray | None
) -> list[int]:
    """Greedy initialisation per Kaufman & Rousseeuw §2.2.

    First medoid m_0 minimises Σ_i w_i · D[i, m] over all candidates m.
    Each subsequent medoid maximises the (weighted) sum of positive
    cost reductions across all points.
    """
    N, M = D.shape
    w = np.ones(N, dtype=np.float64) if weights is None else np.asarray(weights, dtype=np.float64)

    # m_0: candidate column with the smallest weighted-sum distance.
    col_costs = (D * w[:, None]).sum(axis=0)
    first = int(np.argmin(col_costs))
    medoids = [first]
    # nearest_dist[i] = D[i, current_nearest_medoid]
    nearest_dist = D[:, first].copy()

    for _ in range(1, k):
        # Improvement[c] = Σ_i w_i · max(0, nearest_dist[i] - D[i, c]).
        # Highest-improvement candidate becomes the next medoid.
        gain = np.maximum(0.0, nearest_dist[:, None] - D) * w[:, None]
        # Mask out candidates already chosen so we don't pick a duplicate.
        improvements = gain.sum(axis=0)
        improvements[medoids] = -np.inf
        nxt = int(np.argmax(improvements))
        medoids.append(nxt)
        nearest_dist = np.minimum(nearest_dist, D[:, nxt])

    return medoids


def _swap_phase(
    D: np.ndarray,
    medoids: list[int],
    weights: np.ndarray | None,
    max_swaps: int,
) -> list[int]:
    """Local-search refinement per Kaufman & Rousseeuw §2.3.

    Try every (medoid → non-medoid) swap; apply the best improving one;
    repeat until no swap reduces cost or `max_swaps` is reached. This
    is the classical PAM swap — not FasterPAM (Schubert & Rousseeuw
    2019), which would give O(N·M) per iteration vs O(k(M-k)·N) here.
    For our palette sizes (M ≤ 304, k ≤ 128, N typically ≤ 100k) the
    classical version converges in well under a second.
    """
    N, M = D.shape
    w = np.ones(N, dtype=np.float64) if weights is None else np.asarray(weights, dtype=np.float64)
    medoids = list(medoids)

    for _ in range(max_swaps):
        current_cost = _total_cost(D, np.array(medoids), w)
        non_medoids = [c for c in range(M) if c not in medoids]

        best_delta = 0.0
        best_swap: tuple[int, int] | None = None  # (medoid_idx, new_candidate)

        for mi in range(len(medoids)):
            trial = list(medoids)
            for c in non_medoids:
                trial[mi] = c
                cost = _total_cost(D, np.array(trial), w)
                delta = cost - current_cost
                if delta < best_delta:
                    best_delta = delta
                    best_swap = (mi, c)

        if best_swap is None:
            break
        medoids[best_swap[0]] = best_swap[1]

    return medoids


def pam_select_medoids(
    D: np.ndarray,
    k: int,
    weights: np.ndarray | None = None,
    max_swaps: int = 100,
) -> np.ndarray:
    """Pick `k` medoid indices from `D.shape[1]` candidates.

    Args:
      D:        (N, M) precomputed distance matrix. `D[i, j]` is the
                distance from point `i` (e.g. one cell mean) to
                candidate `j` (e.g. one palette chip). The metric is
                the caller's choice — OKLab squared-Euclidean,
                CIEDE2000, etc.
      k:        Number of medoids to select. Must satisfy `0 < k <= M`.
      weights:  Optional (N,) weights. Use cell-frequency counts for
                weighted clustering (saves running PAM over duplicated
                rows). When `None`, all points weight equally.
      max_swaps: Cap on swap-phase iterations.

    Returns:
      (k,) int64 array of indices into the candidate axis of `D`. The
      result is deterministic for a given `D` (no random init — BUILD
      breaks ties by `np.argmin`, which returns the lowest index, and
      SWAP picks the first strictly-improving swap when costs tie).
    """
    D = np.asarray(D, dtype=np.float64)
    if D.ndim != 2:
        raise ValueError(f"D must be 2-D (N, M); got shape {D.shape}")
    N, M = D.shape
    if not 0 < k <= M:
        raise ValueError(f"k must satisfy 0 < k <= M ({M}); got {k}")
    if weights is not None:
        w = np.asarray(weights, dtype=np.float64)
        if w.shape != (N,):
            raise ValueError(f"weights must have shape ({N},); got {w.shape}")

    medoids = _build_phase(D, k, weights)
    medoids = _swap_phase(D, medoids, weights, max_swaps)
    return np.array(sorted(medoids), dtype=np.int64)
